parquet-column/src/main/java/org/apache/parquet/schema/PrimitiveComparator.java

```diff
@@ -0,0 +1,247 @@
+    }
+
+    @Override
+    public String toString() {
+      return "UNSIGNED_INT64_COMPARATOR";
+    }
+  };
+
+  static final PrimitiveComparator<Float> FLOAT_COMPARATOR = new PrimitiveComparator<Float>() {
+    @Override
+    public int compare(Float o1, Float o2) {
+      return compare(o1.floatValue(), o2.floatValue());
+    }
+
+    @Override
+    public int compare(float f1, float f2) {
+      return Float.compare(f1, f2);
+    }
+
+    @Override
+    public String toString() {
+      return "FLOAT_COMPARATOR";
+    }
+  };
+
+  static final PrimitiveComparator<Double> DOUBLE_COMPARATOR = new PrimitiveComparator<Double>() {
+    @Override
+    public int compare(Double o1, Double o2) {
+      return compare(o1.doubleValue(), o2.doubleValue());
+    }
+
+    @Override
+    public int compare(double d1, double d2) {
+      return Double.compare(d1, d2);
+    }
+
+    @Override
+    public String toString() {
+      return "DOUBLE_COMPARATOR";
+    }
+  };
+
+  private static abstract class BinaryComparator extends PrimitiveComparator<Binary> {
+    @Override
+    public final int compare(Binary o1, Binary o2) {
+      return compare(o1.toByteBuffer(), o2.toByteBuffer());
+    }
+
+    abstract int compare(ByteBuffer b1, ByteBuffer b2);
+
+    final int toUnsigned(byte b) {
+      return b & 0xFF;
+    }
+  }
+
+  public static final PrimitiveComparator<Binary> LEXICOGRAPHICAL_BINARY_COMPARATOR = new BinaryComparator() {
+    @Override
+    int compare(ByteBuffer b1, ByteBuffer b2) {
+      int l1 = b1.remaining();
+      int l2 = b2.remaining();
+      int p1 = b1.position();
+      int p2 = b2.position();
+      int minL = Math.min(l1, l2);
+
+      for (int i = 0; i < minL; ++i) {
+        int result = unsignedCompare(b1.get(p1 + i), b2.get(p2 + i));
+        if (result != 0) {
+          return result;
+        }
+      }
+
+      return l1 - l2;
+    }
+
+    private int unsignedCompare(byte b1, byte b2) {
+      return toUnsigned(b1) - toUnsigned(b2);
+    }
+
+    @Override
+    public String toString() {
+      return "LEXICOGRAPHICAL_BINARY_COMPARATOR";
+    }
+  };
+
+  static final PrimitiveComparator<Binary> BINARY_AS_SIGNED_INTEGER_COMPARATOR = new BinaryComparator() {
+    private static final int NEGATIVE_PREFIX = 0xFF;
+    private static final int POSITIVE_PREFIX = 0;
+
+    @Override
+    int compare(ByteBuffer b1, ByteBuffer b2) {
+      int l1 = b1.remaining();
+      int l2 = b2.remaining();
+      int p1 = b1.position();
+      int p2 = b2.position();
+
+      boolean isNegative1 = l1 > 0 ? b1.get(p1) < 0 : false;
+      boolean isNegative2 = l2 > 0 ? b2.get(p2) < 0 : false;
+      if (isNegative1 != isNegative2) {
+        return isNegative1 ? -1 : 1;
+      }
+
+      int maxL = Math.max(l1, l2);
+      int iDiff1 = maxL - l1;
+      int iDiff2 = maxL - l2;
+      int prefix = isNegative1 ? NEGATIVE_PREFIX : POSITIVE_PREFIX;
+      for (int i = 0; i < maxL; ++i) {
```

Comment:
It may be better (and slightly more performant) to implement this as a function with one-sided padding: if `l1 < l2`, return `compareInternal(l1, l2)`, otherwise return `0 - compareInternal(l2, l1)` and always pad `b1` in `compareInternal`. You could also have a third branch for zero padding for decimals backed by fixed-length buffers.

That would eliminate one of the prefix conditionals and would be a bit more clear to someone reading this code.

Reply:
In this suggested implementation I would suggest naming the parameters `shorter` and `longer` instead of `l1` and `l2`.

Reply:
I've chosen a different approach. I'm curious about your opinions.

Reply:
It works.